src/pgrepr/src/value/interval.rs

```diff
@@ -38,10 +37,14 @@ impl ToSql for Interval {
         //
         // Postgres implementation: https://github.com/postgres/postgres/blob/517bf2d91/src/backend/utils/adt/timestamp.c#L1008
         // Diesel implementation: https://github.com/diesel-rs/diesel/blob/a8b52bd05/diesel/src/pg/types/date_and_time/mod.rs#L39
-        out.put_i64(
-            (self.0.duration.as_micros() as i64) * if self.0.is_positive_dur { 1 } else { -1 },
-        );
-        out.put_i32(0);
+        let days = self.0.days() as i128;
+        let sub_day_ns = self.0.duration % (24 * 60 * 60 * 1_000_000_000);
+
+        out.put_i64((sub_day_ns / 1000) as i64);
+        // Postgres' max days is 106,751,991,167,300, which is exactly `i64::MAX
+        // / (60 * 60 * 24)`. Because we avail to keep `Interval` seconds within
+        // i64, this shouldn't compress values.
+        out.put_i32(days as i32);
```

Comment:
@benesch I had a momentary lapse of sense and overlooked that the same problem that affects the Avro encoding also affects the PG encoding; `i32` days is nowhere near sufficient to express the max number of days that PG allows. What do you think we should do here?

Reply:
Note here that I mis-referenced this twice. I stated that PG's max days is > `i32`; that isn't true. _Our_ max days is > `i32`. PG's is, as makes perfect sense, == `i32:::MAX`.

Reply:
I'm not sure I understand how we manage to get away with not having a separate days field, like Postgres? So my mental model here might be way off. Do we diverge on semantics?

Reply:
Not having a day field doesn't meaningfully differentiate our behavior from PG. The only notable difference is that we display hours in excess of 24 as days, where PG does not:

```
pg=# select interval '25 hours';
 interval
----------
 25:00:00
```
```
mz=> select interval '25 hours';
    ?column?
----------------
 1 day 01:00:00
```

PG stores the day in a separate int than the hour:minute:seconds, but rolls over every 24 hours into a day, which means its calculations and ours are identical:

```
pg=# select timestamp '2001-01-01 02:03:04' +  interval '1 day 2147483647 hours';
       ?column?
-----------------------
 246984-10-10 09:03:04
```
```
mz=> select timestamp '2001-01-01 02:03:04' +  interval '1 day 2147483647 hours';
        ?column?
------------------------
 246984-10-10 09:03:04
````

Really, it just seems like Postgres' differentiation of day from the smaller fields is just a matter of compression for them; there's no semantic meaning in that separation like there is for months.

Reply:
Ah got it. 128-bit side types aren’t common in C, and probably didn’t exist at all when PostgreSQL was written. Can we just clamp our max interval to the Postgres max interval? Would that solve the issue?

Reply:
We should probably establish some philosophy at some point about whether our goal is to inherit all of Postgres' limitations, or to build something that could be better / as good as more modern systems. It sort of feels like if you wrote a query that returns a large but totally reasonable interval, and when converting to types pgwire understands it doesn't fit, that moment is where the error occurs?

Reply:
I actually thought we were all aligned on the philosophy of doing things the Postgres way unless there are very clear and specific benefits to deviating? Almost every time we've deviated I've regretted it. Even our different decimal implementation causes a lot of headaches, although I think the justification for that one is sound (in-place differential aggregates). If you have concerns about this philosophy, Frank, perhaps we should try to sync up with product and nail down something more concrete?

> It sort of feels like if you wrote a query that returns a large but totally reasonable interval, and when converting to types pgwire understands it doesn't fit, that moment is where the error occurs?

Strictly speaking, yes, though having data we can represent internally but not via pgwire definitely feels a bit weird, considering pgwire is the only "mature" interface to Materialize today.

In this specific case, the maximum interval that Postgres supports is pretty darn massive. Most programming languages I'm aware of represent durations as a type with less range than Postgres, so users don't typically feel the pain.

Perhaps a way of looking at this that is more palatable is that we are choosing types that are likely to be immediately useful to users in other programming languages? There are existing Postgres drivers that know how to take an interval in the Postgres range and map it to something sensible in that driver's language. If we invent a better interval type, we are now on the hook for writing that code in every language that people want to use Materialize with, or at least telling folks how to perform the conversion.

Down the road when we have more resources, I expect that there may be types where we want to provide a Materialize variant, like perhaps a `MaterializeInterval` and a `PostgresInterval`. Sophisticated users in the know could then opt-in to the new types, and write the necessary support glue for their drivers, while everyone else could use the Postgres compatible types.

Reply:
Made newly-created Intervals PG compatible; still need to propagate this change across all operators, but this shows it in action. Resolving this convo, but glad to revisit if anyone has unvoiced concerns.